### PR TITLE
[basicprofiles] Add inequality comparisons to State Filter profile

### PR DESCRIPTION
**Pull request by jimtng**
- Add inequality comparisons to State Filter profile
- Support symbolic operators `==`, `!=`, etc.
- Fix a bug where undefined `mismatchState` passed `UNDEF` instead of ignoring state updates
- Support multiline conditions
- Support comparing against the input state from handler to filter out
unwanted data
- Support comparing the states of item1 vs item2
- Support comparing the input state against another item's state

https://community.openhab.org/t/state-filter-range-filter-profile/158025

### CODE REVIEW

**Comment by jlaur** [comment]
@J-N-K - would you like to add yourself to `CODEOWNERS` for **basicprofiles**? I missed this in #16754.

File: bundles/org.openhab.transform.basicprofiles/README.md [hunk 1]
```diff
@@ -27,13 +27,13 @@ Switch lightsStatus {
 
 The Generic Toggle Switch Profile is a specialization of the Generic Command Profile and toggles the State of a Switch Item whenever one of the specified events is triggered.
 
-### Configuration
+### Generic Toggle Switch Profile Configuration
```
Comment (J-N-K) [inline bundles/org.openhab.transform.basicprofiles/README.md:30]: Why did you change that? Isn't this already clear by the structure (`##` denotes the profile, `###` the section within that profile)

File: bundles/org.openhab.transform.basicprofiles/README.md [hunk 2]
```diff
@@ -180,31 +180,81 @@ Switch motionSensorFirstFloor {
...
+| `GTE` |  `>=`  | Greater than or equals to |
+| `LT`  |  `<`   | Less than                 |
+| `LTE` |  `<=`  | Less than or equals to    |
 
-### Full Example
+Notes:
+
+- The operator names must be surrounded by spaces, i.e.: `Item EQ 10`
+- The operator symbols do not need to be surrounded by spaces, e.g.: `Item==10` and `Item == 10` are both fine.
+- Only symbolic operators can be used when comparing against the incoming state, e.g. `> 10`. Using operator names isn't supported, i.e. this is not supported: ~~`GT 10`~~.
```
Comment (J-N-K) [inline bundles/org.openhab.transform.basicprofiles/README.md:231]: Why is that?

File: bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java [hunk 3]
```diff
@@ -40,59 +48,75 @@
...
 
-            return parsedConditions;
-        } catch (IllegalArgumentException e) {
-            logger.warn("Cannot parse condition {}. Expected format ITEM_NAME <EQ|NEQ> STATE_VALUE: '{}'", config,
-                    e.getMessage());
-            return List.of();
-        }
+        String operatorPattern = Stream.of(StateCondition.ComparisonType.values())
+                .flatMap(type -> Stream.of(" " + type.name() + " ", type.getSymbol()))
+                .sorted((a, b) -> Integer.compare(b.length(), a.length())) // We want to match the longest operator
```
Comment (J-N-K) [inline bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java:86]:
```suggestion
                .sorted(Comparator.comparingInt(String::length)) // We want to match the longest operator
```

File: bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java [hunk 4]
```diff
@@ -40,59 +48,75 @@
...
-                }
-            }
 
-            return parsedConditions;
-        } catch (IllegalArgumentException e) {
-            logger.warn("Cannot parse condition {}. Expected format ITEM_NAME <EQ|NEQ> STATE_VALUE: '{}'", config,
-                    e.getMessage());
-            return List.of();
-        }
+        String operatorPattern = Stream.of(StateCondition.ComparisonType.values())
```
Comment (J-N-K) [inline bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java:84]: You could make that (and the resulting `Pattern`) a static constant and improve performance.

File: bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java [hunk 5]
```diff
@@ -173,47 +180,177 @@ State parseState(@Nullable String stateString) {
...
+                            return true;
+                        } else {
+                            logger.debug("Condition value: '{}' is not compatible with state '{}' ({})", value, state,
+                                    state.getClass().getSimpleName());
+                            return false;
+                        }
+                    }
+                }
+
+                rhs = Objects.requireNonNull(parsedValue instanceof StringType ? parsedValue.toString() : parsedValue);
```
Comment (J-N-K) [inline bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java:284]: Why do you need `Objects.requireNonNull` here? If `parsedValue==null` you exit in the if-clause above.

File: bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java [hunk 6]
```diff
@@ -173,47 +180,177 @@ State parseState(@Nullable String stateString) {
...
+                logger.warn("Error evaluating condition: {}: {}", this, e.getMessage());
             }
+            return false;
         }
 
         enum ComparisonType {
-            EQ,
-            NEQ
+            EQ("=="),
+            NEQ("!="),
```
Comment (J-N-K) [inline bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java:312]: What about `<>`? IMO this is not uncommon for "not equal". Maybe you can add `NEQ_ALT("<>")` and also add `NEQ_ALT` to the `NEQ` case above.

File: bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java [hunk 7]
```diff
@@ -173,47 +180,177 @@ State parseState(@Nullable String stateString) {
...
+
+            ComparisonType(String symbol) {
+                this.symbol = symbol;
+            }
+
+            String getSymbol() {
+                return symbol;
+            }
+
+            static @Nullable ComparisonType fromSymbol(String symbol) {
```
Comment (J-N-K) [inline bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java:328]: If you use `Optional<ComparisonType>` as return value, you can use `fromSymbol(...).orElse(...)` in line 110 instead of a null-check.

**Review by J-N-K — CHANGES_REQUESTED** [review]
In general: nice addition.

File: bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java (hunk 3 above)
Comment (jimtng) [reply bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java:86]: Thanks for the tip

File: bundles/org.openhab.transform.basicprofiles/README.md (hunk 2 above)
Comment (jimtng) [reply bundles/org.openhab.transform.basicprofiles/README.md:231]: The regex is more complicated to support it, but I've redone it now.

File: bundles/org.openhab.transform.basicprofiles/README.md (hunk 1 above)
Comment (jimtng) [reply bundles/org.openhab.transform.basicprofiles/README.md:30]: 1. I followed the markdown warning that the headings should be unique.
2. This makes the anchor links better, but it's a very obscure benefit perhaps.  Without this, the links would look like `#configuration-1`, `#configuration-2`, etc.

Should I revert this?

File: bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java (hunk 5 above)
Comment (jimtng) [reply bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java:295]: The compiler complained about null mismatch further down in the calls to `compareTo(rhs)`.

File: bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java (hunk 6 above)
Comment (jimtng) [reply bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java:323]: I've added `<>`

File: bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java (hunk 7 above)
Comment (jimtng) [reply bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java]: done

File: bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java (hunk 4 above)
Comment (jimtng) [reply bundles/org.openhab.transform.basicprofiles/src/main/java/org/openhab/transform/basicprofiles/internal/profiles/StateFilterProfile.java]: done
